src/osd/OSD.cc

```diff
@@ -3278,6 +3278,13 @@ will start to track new ops received afterwards.";
       st.dump(f);
       f->close_section();
     }
+  } else if (prefix == "dump_osd_pg_stats") {
+    lock_guard l(osd_lock);
+
+    MPGStats* m = collect_pg_stats();
+    ceph_assert(m);
+    m->dump_stats(f);
+    MessageRef mref(m); // to trigger m's release
```

Comment:
is this the best way for that?

Reply:
Agree this is a bit ugly but I don't see anything simpler... 
~MPGStats is declared as private so we either need to modify it or retrieve the stats via some replacement structure/method.

Reply:
Your call. But looking at the existing client of collect_pg_stats(),
as an mgrc callback, I wonder how the release is handled there. For one thing - that
code is far from clear when only looking at the h files.

So - maybe unrelated to this PR, we might want to modify the returned type for collect_pg_stats().

Reply:
> Your call. But looking at the existing client of collect_pg_stats(), as an mgrc callback, I wonder how the release is handled there. For one thing - that code is far from clear when only looking at the h files.

IMO this works fine - Message* goes to AsyncConnection::send_message() and further to ProtocolV1/ProcotolV2::send_message() funcs which care about the release through m->put()

The latter is a proper way to release such a pointer indeed. Updated my patch...

> 
> So - maybe unrelated to this PR, we might want to modify the returned type for collect_pg_stats().
> 
> NB - shouldn't m->put() work?
yep, this is a bit nicer.